lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1417 @@
+    $posixIpcEgg = Join-Path $LIB_DIR "posix_ipc-0.9.8-py2.7.egg-info"
+    if (!(Test-Path $posixIpcEgg)) {
+        Expand-ZipArchive $zipPath $LIB_DIR
+    }
+}
+
+
+function Install-PyWin32 {
+    Write-JujuLog "Installing pywin32"
+    Start-ExternalCommand -ScriptBlock { pip install pywin32 } `
+                          -ErrorMessage "Failed to install pywin32."
+    Start-ExternalCommand {
+        python "$PYTHON_DIR\Scripts\pywin32_postinstall.py" -install
+    } -ErrorMessage "Failed to run pywin32_postinstall.py"
+}
+
+
+function Install-FreeRDPConsole {
+    Write-JujuLog "Installing FreeRDP"
+
+    Install-VC2012
+
+    $freeRDPZip = Join-Path $FILES_DIR "FreeRDP_powershell.zip"
+    $charmLibDir = Join-Path (Get-JujuCharmDir) "lib"
+    Expand-ZipArchive $freeRDPZip $charmLibDir
+
+    # Copy wfreerdp.exe and DLL file to Windows folder
+    $freeRDPFiles = @('wfreerdp.exe', 'libeay32.dll', 'ssleay32.dll')
+    $windows = Join-Path $env:SystemDrive "Windows"
+    foreach ($file in $freeRDPFiles) {
+        Copy-Item "$charmLibDir\FreeRDP\$file" $windows
+    }
+
+    $freeRDPModuleFolder = Join-Path $windows "system32\WindowsPowerShell\v1.0\Modules\FreeRDP"
+    if (!(Test-Path $freeRDPModuleFolder)) {
+        mkdir $freeRDPModuleFolder
+    }
+    Copy-Item "$charmLibDir\FreeRDP\FreeRDP.psm1" $freeRDPModuleFolder
+    Remove-Item -Recurse "$charmLibDir\FreeRDP"
+
+    Write-JujuLog "Finished installing FreeRDP."
+}
+
+
+function Write-PipConfigFile {
+    $pipDir = Join-Path $env:APPDATA "pip"
+    if (Test-Path $pipDir){
+        Remove-Item -Force -Recurse $pipDir
+    }
+
+    $pypiMirror = Get-JujuCharmConfig -scope 'pypi-mirror'
+    if ($pypiMirror -eq $null -or $pypiMirror.Length -eq 0) {
+        Write-JujuLog ("pypi-mirror config is not present. " +
+                       "Will not generate the pip.ini file.")
+        return
+    }
+    mkdir $pipDir
+    $pipIni = Join-Path $pipDir "pip.ini"
+    New-Item -ItemType File $pipIni
+
+    $mirrors = $pypiMirror.Split()
+    $hosts = @()
+    foreach ($i in $mirrors){
+        $h = ([System.Uri]$i).Host
+        if ($h -in $hosts) {
+            continue
+        }
+        $hosts += $h
+    }
+
+    Set-IniFileValue "index-url" "global" $mirrors[0] $pipIni
+    if ($mirrors.Length -gt 1){
+        Set-IniFileValue "extra-index-url" "global" ($mirrors[1..$mirrors.Length] -Join " ") $pipIni
+    }
+    Set-IniFileValue "trusted-host" "install" ($hosts -Join " ") $pipIni
+}
+
+
+function Get-HypervADUser {
+    $adUsername = Get-JujuCharmConfig -scope 'ad-user-name'
+    if (!$adUsername) {
+        $adUsername = "hyper-v-user"
+    }
+    return $adUsername
+}
+
+
+function Set-DevStackRelationParams {
+    Param(
+        [HashTable]$RelationParams
+    )
+
+    $rids = Get-JujuRelationIds -Relation "devstack"
+    foreach ($rid in $rids) {
+        try {
+            Set-JujuRelation -Settings $RelationParams -RelationId $rid
```

Comment:
Don't wrap this in a `try { } catch { }` as we want it to be fatal if it doesn't succeed. 

And since you are using this function only in the `Set-AD2DevstackCreds` (which is short as well), I would recommend moving this logic there (the body of this function is just a `for`).